include/yk/hash/hash_value_for.hpp

```diff
@@ -0,0 +1,32 @@
+#ifndef YK_HASH_HASH_VALUE_FOR_HPP
+#define YK_HASH_HASH_VALUE_FOR_HPP
+
+#include "yk/hash/hash_value_for/std.hpp"
+
+#if YK_UTIL_INCLUDE_BOOST
+#include "yk/hash/hash_value_for/boost.hpp"
+#include "yk/hash/range.hpp"
+#endif
+
+namespace yk {
+
+template <class T>
+[[nodiscard]] constexpr std::size_t hash_value_for(const T& x) noexcept {
+  if constexpr (requires(std::hash<T> hasher) { hasher(x); }) {
+    return std::hash<T>{}(x);
+  } else if constexpr (requires { hash_value(x); }) {
+    return hash_value(x);
+  }
+#if YK_UTIL_INCLUDE_BOOST
+  else if constexpr (requires { ::yk::hash_range(x); }) {
+    return ::yk::hash_range(x);
+  }
+#endif
```

Comment:
This leads to ODR violation.
I don't see much benefit on removing boost dependency here; we can just use the boost hashers since we do need them.

Reply:
`hash_value_for<T>` where `T` satisfies `requires { ::yk::hash_range(x); }` is only instantiated if the condition is met, so I think this does not lead to ODR violation, but undefined reference.

Reply:
This leads to ODR violation

a.cpp
```cpp
#define YK_UTIL_INCLUDE_BOOST 1
#include <yk/hash/hash_value_for.hpp>
auto aval = yk::hash_value_for({1, 2, 3});
```

b.cpp
```cpp
#define YK_UTIL_INCLUDE_BOOST 0
#include <yk/hash/hash_value_for.hpp>
auto bval = yk::hash_value_for({1, 2, 3});
```

main.exe: a.cpp b.cpp


Also, ODR violation has nothing to do with whether the template is instantiated or not; it's all about *definition* and when there exists two conflicting definitions for odr-used entity, it will lead to ODR violation.

Reply:
Another example

```cpp
void func()
{
#if BOOST_MSVC
    do_msvc_thing();
#endif
    do_process();
}
```

The more commonly used pattern like above would not lead to ODR violation because the macro evaluation result is guaranteed to be consistent on a single configuration, no magic is involved.